alchemy/alchemy.py

```diff
@@ -971,40 +1168,82 @@ def _alchemicallyModifyNonbondedForce(self, system, reference_force):
-                # Zero terms in NonbondedForce.
-                nonbonded_force.setExceptionParameters(exception_index, iatom, jatom, abs(0*chargeprod), sigma, abs(0*epsilon))
-
-        # TODO: Add back NonbondedForce terms for alchemical system needed in case of decoupling electrostatics or sterics via second CustomBondForce.
-        # TODO: Also need to change current CustomBondForce to not alchemically disappearing system.
-
-        # Restrict interaction evaluation to be between alchemical atoms and rest of environment.
-        # TODO: Exclude intra-alchemical region if we are separately handling that through a separate CustomNonbondedForce for decoupling.
-        sterics_custom_nonbonded_force.addInteractionGroup(list(atomset1), list(atomset2))
-        electrostatics_custom_nonbonded_force.addInteractionGroup(list(atomset1), list(atomset2))
+
+            # Exclude this atom pair in CustomNonbondedForces. All nonbonded forces
+            # must have the same number of exceptions/exclusions on CUDA platform.
+            na_sterics_custom_nonbonded_force.addExclusion(iatom, jatom)
+            aa_sterics_custom_nonbonded_force.addExclusion(iatom, jatom)
+            na_electrostatics_custom_nonbonded_force.addExclusion(iatom, jatom)
+            aa_electrostatics_custom_nonbonded_force.addExclusion(iatom, jatom)
+
+            # Check how many alchemical atoms we have
+            both_alchemical = iatom in alchemical_atomset and jatom in alchemical_atomset
+            only_one_alchemical = (iatom in alchemical_atomset) != (jatom in alchemical_atomset)
+
+            # Check if this is an exception or an exclusion
+            is_exception_epsilon = abs(epsilon.value_in_unit_system(unit.md_unit_system)) > 0.0
+            is_exception_chargeprod = abs(chargeprod.value_in_unit_system(unit.md_unit_system)) > 0.0
+
+            # If exception (and not exclusion), add special CustomBondForce terms to
+            # handle alchemically-modified Lennard-Jones and electrostatics exceptions
+            if both_alchemical:
+                if is_exception_epsilon:
+                    aa_sterics_custom_bond_force.addBond(iatom, jatom, [sigma, epsilon])
+                if is_exception_chargeprod:
+                    aa_electrostatics_custom_bond_force.addBond(iatom, jatom, [chargeprod, sigma])
+            elif only_one_alchemical:
+                if is_exception_epsilon:
+                    na_sterics_custom_bond_force.addBond(iatom, jatom, [sigma, epsilon])
+                if is_exception_chargeprod:
+                    na_electrostatics_custom_bond_force.addBond(iatom, jatom, [chargeprod, sigma])
+            # else: both particles are non-alchemical, leave them in the unmodified NonbondedForce
+
+        # Turn off all exception contributions from alchemical atoms in the NonbondedForce
+        # modelling non-alchemical atoms only
+        for exception_index in range(nonbonded_force.getNumExceptions()):
```

Comment:
I purposely left it on a second loop because I thought it would make it more readable, but I can include it in the previous for loop if you think it's better.